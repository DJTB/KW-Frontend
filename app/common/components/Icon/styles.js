import styled, { css } from 'styled-components';
import { transparent } from 'common/styles/colors';
import { spin } from 'common/styles/animation';

export const SVGWrapper = styled.span`
  ${({ inline }) => inline ? css`
    display: inline-block;
    vertical-align: middle;
  ` : css`
    display: block;
    align-self: center;
  `}
  position: relative;
  width: ${({ size }) => size}; /*CSS instead of html width attr to support non-pixel units*/
  height: ${({ size }) => size}; /*Prevents scaling issue in IE*/
  background-repeat: no-repeat;
  background-color: ${transparent};
  color: ${({ color }) => color};
  flex-shrink: 0;
  ${({ isRotating }) => isRotating && css`animation: ${spin} 1.25s linear infinite;`};
`;

export const SVG = styled.svg`
  display: block;
  pointer-events: none;
  transform-origin: 50% 50% 0px;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  fill: currentColor;
`;
