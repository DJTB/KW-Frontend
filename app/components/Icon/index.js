import React, { PropTypes } from 'react';
import styled from 'styled-components';
import ReactTooltip from 'react-tooltip';
import cuid from 'cuid';
import { ICONS } from './constants';

const Wrapper = styled.span`
  display: inline-block;
  vertical-align: middle;
  align-self: center;
  position: relative;
  width: ${({ iconSize }) => iconSize}; /*CSS instead of html width attr to support non-pixel units*/
  height: ${({ iconSize }) => iconSize}; /*Prevents scaling issue in IE*/
  background-repeat: no-repeat;
  transition: all 200ms ease-in-out;
  color: ${({ iconColor }) => iconColor};
`;

const SVG = styled.svg`
  display: block;
  pointer-events: none;
  transform-origin: 50% 50% 0px;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
`;

const tooltipDefaults = {
  position: 'right',
  showDelay: 0,
  hideDelay: 0,
};

const Icon = ({ color, size, name, className, tooltip, ...rest }) => {
  // NOTE: <ReactTooltip /> must be present in a parent component (pref root) for tooltips to show!
  const tooltipOptions = Object.assign(
    {},
    tooltipDefaults,
    tooltip,
    { id: cuid() },
  );
  return (
    <Wrapper
      className={className}
      iconColor={color}
      iconSize={size}
      data-tip={tooltipOptions.text}
      data-for={tooltipOptions.id}
      data-place={tooltipOptions.position}
      data-delay-show={tooltipOptions.showDelay}
      data-delay-hide={tooltipOptions.hideDelay}
    >
      { tooltipOptions.text && <ReactTooltip id={tooltipOptions.id} /> }
      <SVG
        title={name}
        width="100%"
        height="100%"
        {...rest}
      >
        <path d={ICONS[name]} />
      </SVG>
    </Wrapper>
  );
};

Icon.propTypes = {
  name: PropTypes.oneOf(Object.keys(ICONS)).isRequired,
  className: PropTypes.string,
  color: PropTypes.string,
  size: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  viewBox: PropTypes.string,
  preserveAspectRatio: PropTypes.string,
  tooltip: PropTypes.shape({
    text: PropTypes.string.isRequired,
    position: PropTypes.string,
    showDelay: PropTypes.number,
    hideDelay: PropTypes.number,
  }),
};

Icon.defaultProps = {
  color: 'currentColor',
  size: '1em',
  viewBox: '0 0 24 24', // polymer default
  preserveAspectRatio: 'xMidYMid meet',
};

export default Icon;
