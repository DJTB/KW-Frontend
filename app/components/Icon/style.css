.wrapper {
    display: inline-block;
    vertical-align: middle;
    position: relative;
    background-repeat: no-repeat;
    transition: all 250ms ease-in-out;
}
.svg {
    display: block;
    pointer-events: none; /* capture bubbled events on wrapper instead */
    transform-origin: 50% 50% 0px;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
}
