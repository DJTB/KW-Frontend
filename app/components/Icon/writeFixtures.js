import fs from 'fs';
import { ICONS } from './constants';

Object.keys(ICONS).forEach((name) => {
  const fileName = `__fixtures__/${name.toLowerCase()}.js`;
  const content = `
    export default {
      props: {
        name: '${name}',
        size: '5rem',
      },
    };`;
  fs.writeFile(fileName, content, (err) => err ? console.error(err) : console.log(`${fileName} generated`));
});
