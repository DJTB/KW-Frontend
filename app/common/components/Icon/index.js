import React from 'react';
import PropTypes from 'prop-types';

import ICONS from './constants';
import { SVGWrapper, SVG } from './styles';

Icon.propTypes = {
  name: PropTypes.oneOf(Object.keys(ICONS)).isRequired,
  inline: PropTypes.bool,
  color: PropTypes.string,
  size: PropTypes.string,
  isRotating: PropTypes.bool,
};

Icon.defaultProps = {
  inline: true,
  color: 'currentColor',
  size: '1.5em',
  isRotating: false,
};

function Icon({ name, ...props }) {
  const NAME = ICONS[name] || {};
  return (
    <SVGWrapper {...props}>
      <SVG
        title={name}
        width="100%"
        height="100%"
        preserveAspectRatio="xMidYMid meet"
        viewBox={NAME.viewBox}
      >
        <path d={NAME.path} />
      </SVG>
    </SVGWrapper>
  );
}

export default Icon;
