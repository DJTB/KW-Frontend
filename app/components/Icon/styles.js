import styled from 'styled-components';
import { transparent } from 'shared/styles/colors';

export const SVGWrapper = styled.span`
  ${({ inline }) => inline ? `
    display: inline-block;
    vertical-align: middle;
  ` : `
    display: block;
    align-self: center;
  `}
  position: relative;
  width: ${({ size }) => size}; /*CSS instead of html width attr to support non-pixel units*/
  height: ${({ size }) => size}; /*Prevents scaling issue in IE*/
  background-repeat: no-repeat;
  background-color: ${transparent};
  color: ${({ color }) => color};
`;

export const SVG = styled.svg`
  display: block;
  pointer-events: none;
  transform-origin: 50% 50% 0px;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  fill: currentColor;
`;
