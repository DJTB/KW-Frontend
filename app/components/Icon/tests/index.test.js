import 'jest-styled-components';
import React from 'react';
import { shallow } from 'enzyme';
import Icon from '../index';

describe('<Icon /> ', () => {
  it('should match baseline snapshot', () => {
    const renderedComponent = shallow(<Icon name="ADD" />);
    expect(renderedComponent).toMatchSnapshot();
  });
});
