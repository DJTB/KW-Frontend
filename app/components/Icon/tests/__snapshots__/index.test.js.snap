// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon />  should match baseline snapshot 1`] = `
<styled.span
  color="currentColor"
  display="inline-block"
  size="1.5em"
>
  <styled.svg
    height="100%"
    preserveAspectRatio="xMidYMid meet"
    title="ADD"
    viewBox="0 0 24 24"
    width="100%"
  >
    <path
      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
    />
  </styled.svg>
</styled.span>
`;
